mobile/lib/domain/services/sync_stream.service.dart

```diff
@@ -42,180 +28,97 @@ class SyncStreamService {
+      case SyncEntityType.assetV1:
+        return _syncStreamRepository.updateAssetsV1(data.cast());
+      case SyncEntityType.assetDeleteV1:
+        return _syncStreamRepository.deleteAssetsV1(data.cast());
+      case SyncEntityType.assetExifV1:
+        return _syncStreamRepository.updateAssetsExifV1(data.cast());
+      case SyncEntityType.partnerAssetV1:
+        return _syncStreamRepository.updatePartnerAssetsV1(data.cast());
+      case SyncEntityType.partnerAssetDeleteV1:
+        return _syncStreamRepository.deletePartnerAssetsV1(data.cast());
+      case SyncEntityType.partnerAssetExifV1:
+        return _syncStreamRepository.updatePartnerAssetsExifV1(data.cast());
+      default:
+        _logger.warning("Unknown sync data type: $type");
+        return false;
+    }
+  }
 
+  Future<void> _handleEvents(List<SyncEvent> events, Function() abort) async {
+    List<String> acksToSend = [];
     try {
-      if (type == SyncEntityType.partnerV1) {
-        return await _syncStreamRepository.updatePartnerV1(data.cast());
-      }
-
-      if (type == SyncEntityType.partnerDeleteV1) {
-        return await _syncStreamRepository.deletePartnerV1(data.cast());
-      }
-
-      if (type == SyncEntityType.userV1) {
-        return await _syncStreamRepository.updateUsersV1(data.cast());
-      }
-
-      if (type == SyncEntityType.userDeleteV1) {
-        return await _syncStreamRepository.deleteUsersV1(data.cast());
-      }
-
-      if (type == SyncEntityType.assetV1) {
-        return await _syncStreamRepository.updateAssetsV1(data.cast());
-      }
-
-      if (type == SyncEntityType.assetDeleteV1) {
-        return await _syncStreamRepository.deleteAssetsV1(data.cast());
-      }
-
-      if (type == SyncEntityType.assetExifV1) {
-        return await _syncStreamRepository.updateAssetsExifV1(data.cast());
-      }
-
-      if (type == SyncEntityType.partnerAssetV1) {
-        return await _syncStreamRepository.updatePartnerAssetsV1(data.cast());
-      }
+      SyncEntityType? currentBatchType;
+      String? currentBatchAck;
+      // ignore: avoid-dynamic
+      List<dynamic> currentBatchData = [];
+
+      for (final event in events) {
+        if (_cancelChecker?.call() ?? false) {
+          _logger.warning("Sync stream cancelled");
+          // Clear the current batch data to avoid processing it
+          currentBatchData.clear();
+          abort();
+          break;
+        }
+
+        if (event.type == currentBatchType) {
+          currentBatchData.add(event.data);
+          currentBatchAck = event.ack;
+          continue;
+        }
+
+        if (currentBatchData.isNotEmpty) {
+          if (await _handleSyncData(currentBatchType!, currentBatchData)) {
+            acksToSend.add(currentBatchAck!);
+          } else {
+            _logger
+                .warning("Failed to process sync data for $currentBatchType");
+            abort();
+            currentBatchData.clear();
+            break;
+          }
+        }
```

Comment:
this same logic exists twice. I would probably just put it in a private method and then call it here and after the loop
```ts
for(...) {
  
  if(changed){
    _onBatch(items, ack);
  }
  ...
}

_onBatch(items, ack);
```

This works really well, especially if you make `_onBatch()` accept undefined/empty lists that just returns without doing anything.